math.hpp

```diff
@@ -113,6 +133,262 @@ template<typename U, typename S, typename V, typename T>
+template<typename V  >                       constexpr auto get_tol_val(tolerance_aux<tolerance_nom_tag,V,V>     n) {return get_tol_val(n, tolerance_traits<V>::fraction());}
+template<typename F, typename Q>             constexpr auto get_tol_val(tolerance_aux<tolerance_frc_tag,F,Q>     f) {return get_tol_val(f, tolerance_traits<Q>::nominal ());}
+
+template<typename T1, typename T2, typename Tol> constexpr bool is_near_impl   (T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return abs(lval-rval) <= abs(tol.v);}
+template<typename T1, typename T2, typename Tol> constexpr bool is_smaller_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near(lval,rval,tol) and lval-rval < abs(tol.v);}
+template<typename T1, typename T2, typename Tol> constexpr bool is_greater_impl(T1 lval, T2 rval, tolerance_val<Tol> tol) {using std::abs; return not is_near(lval,rval,tol) and lval-rval > abs(tol.v);}
+
+template<typename T1, typename T2, typename Aux>                 constexpr detail::if_float2_pt_t<T1,T2,bool> is_near   (T1 lval, T2 rval, Aux  tol)             {return is_near_impl   (lval, rval, get_tol_val(tol));}
+template<typename T1, typename T2, typename Aux>                 constexpr detail::if_float2_pt_t<T1,T2,bool> is_smaller(T1 lval, T2 rval, Aux  tol)             {return is_smaller_impl(lval, rval, get_tol_val(tol));}
+template<typename T1, typename T2, typename Aux>                 constexpr detail::if_float2_pt_t<T1,T2,bool> is_greater(T1 lval, T2 rval, Aux  tol)             {return is_greater_impl(lval, rval, get_tol_val(tol));}
+template<typename T1, typename T2, typename Aux1, typename Aux2> constexpr detail::if_float2_pt_t<T1,T2,bool> is_near   (T1 lval, T2 rval, Aux1 tol1, Aux2 tol2) {return is_near_impl   (lval, rval, get_tol_val(tol1,tol2));}
+template<typename T1, typename T2, typename Aux1, typename Aux2> constexpr detail::if_float2_pt_t<T1,T2,bool> is_smaller(T1 lval, T2 rval, Aux1 tol1, Aux2 tol2) {return is_smaller_impl(lval, rval, get_tol_val(tol1,tol2));}
+template<typename T1, typename T2, typename Aux1, typename Aux2> constexpr detail::if_float2_pt_t<T1,T2,bool> is_greater(T1 lval, T2 rval, Aux1 tol1, Aux2 tol2) {return is_greater_impl(lval, rval, get_tol_val(tol1,tol2));}
+
+}
+
+/** @{
+ * @brief compare two floating point values
+ *
+ * This compares two floating point values, or quantities of floating point
+ * values.
+ *
+ * @param lval              left value to compare
+ * @param rval              right value to compare
+ * @param tol, tol1, tol2   tolerance
+ *
+ * @note As tolerance, either an absolute tolerance value (created via
+ *       @ref tolerance_value()) or a tolerance nominal (created via @ref
+ *       tolerance_nominal()) and a tolerance fraction (created via @ref
+ *       tolerance_fraction()) can be passed. Either the nominal or the
+ *       fraction can also be omitted, in which case the default is obtained
+ *       via the @ref tolerance_traits.
```

Comment:
There is an ambiguity in how the tolerances could be applied to, e.g. `is_smaller`:
- it could "soften" the requirement, so that `is_smaller(1.0, 0.99, 0.1) == true` (not the case here)
- it could "harden" the requirement, so that `is_smaller(0.99, 1.0, 0.1) == false` (how it is done)

I think the fine documentation should specify that.

Also, tolerance boundaries could be inclusive (`is_smaller(a,b,tol) == a <= b-tol`) or exclusive (`is_smaller(a,b,tol) == a < b-tol`). This should be documented as well.

Reply:
I am not quite sure of the semantics of your "soften" and "harden", but the tolerances should be applied so that values that are equal within the given tolerance are not considered smaller or greater, even if they would be without the tolerance. I'll try to make the documentation say so.

And, yes, I did not document that all tolerances are inclusive. Given that we think and calculate in decimal numbers, while the machine does in binary ones, causing lots of inaccuracies, I didn't think this would make much of a difference anyway. But I will document this anyway.